Avoid grabbing DeviceGuard in at::empty when possible

zou3519 opened the pull request:
Changed at::empty to allocate the correct amount of memory instead of
"allocate 0 memory and then resize it to the necessary size".

This leads to a 300 ns speedup for at::empty for a cuda tensor of size (64, 2048).
(1790ns -> 1460ns for at::empty).

Also does the following:
Removes DeviceGuards for:
- empty_* functions that end up calling functions that already have a
  DeviceGuard
- t(), which gets called a lot in LSTMs,
- Remove one of the two DeviceGuard that at::empty(...) uses. It only
  needs one for correctness, the other comes from the resize_
  implementation.

Test Plan:
- run tests

zou3519 committed:
Remove some more unnecessary DeviceGuard

Removes DeviceGuards for:
- empty_* functions that end up calling functions that already have a
  DeviceGuard
- t(), which gets called a lot in LSTMs,
- Remove one of the two DeviceGuard that at::empty(...) uses. It only
  needs one for correctness, the other comes from the resize_
  implementation.

Test Plan:
- code reading

colesbury commented:
This will likely have a merge conflict with https://github.com/pytorch/pytorch/pull/13751

colesbury reviewed (changes requested):
I'm concerned by a few things:

1) The static_cast of the allocator
2) The calls into the CUDA API from a different device than the one allocating

For (2), I'm concerned that a call like `torch.empty(10, device=2)` will still initialize the current device (0) due to the CUDA calls in `process_events()`

ezyang commented:
I just merged the DeviceGuard patch, you probably want to rebase on top of that one.

ezyang commented on aten/src/THC/THCCachingAllocator.cpp:
```diff
@@ -179,6 +180,24 @@ struct THCCachingAllocator
+
+      // NB: This is an optimization. If we know we are allocating on the current
+      // device then there is no need to DeviceGuard.
+      c10::optional<at::cuda::CUDAGuard> guard;
```
Post my DeviceGuard patch, you should write `at::cuda::OptionalCUDAGuard` here now. (DeviceGuard no longer has a default constructor.)

zou3519 committed:
Avoid grabbing DeviceGuard in at::empty when possible

Changed at::empty to allocate the correct amount of memory instead of
"allocate 0 memory and then resize it to the necessary size".

This leads to a 300 ns speedup for at::empty for a cuda tensor of size (64, 2048).
(1790ns -> 1460ns for at::empty).

Test Plan:
- run tests

zou3519 commented:
Discussed with @colesbury and we decided it was better to not complicate the THCCachingAllocator with an "allocate_on_device" function. I've removed that part of the PR; now this PR only removes the use of `resize_` in `at::empty`

zou3519 commented:
Test failures should be fixed by #13864

zou3519 committed:
Fix bug in sum_intlist and prod_intlist

zou3519 committed:
Fix windows tests
